File: downloader/server/src/runtime.js
import { timingSafeEqual } from 'node:crypto';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

// Constant-time secret comparison, matching the Python services' compare_digest: the threat model
// is a hostile local process, which is exactly the attacker able to time loopback replies.
// Length is checked first because timingSafeEqual throws on unequal buffers, and the secret's
// length is fixed by the launcher and not itself a secret.
function secretMatches(given, secret) {
  if (typeof given !== 'string') return false;
  const a = Buffer.from(given);
  const b = Buffer.from(secret);
  return a.length === b.length && timingSafeEqual(a, b);
}

// Binds the express app to loopback only and reports the port it actually got.
// FASTSTUDY_PORT=0 asks the OS for an ephemeral port, so the bound port is read back off
// server.address() and printed alone on its own line for the launcher to parse.
export function serve(app, defaultPort, onListening) {
  const port = Number(process.env.FASTSTUDY_PORT ?? defaultPort);
  const server = app.listen(port, '127.0.0.1', () => {
    const boundPort = server.address().port;
    console.log(`FASTSTUDY_PORT=${boundPort}`);
    onListening(boundPort);
  });
  return server;
}

// Rejects any request that does not carry the launch secret, as an X-FastStudy-Secret header or
// a `secret` query parameter — EventSource is the one caller that cannot set a header.
// Unset FASTSTUDY_SECRET (dev) means no enforcement at all.
export function requireSecret(req, res, next) {
  const secret = process.env.FASTSTUDY_SECRET;
  // /health is exempt by path, not merely by registration order: the launcher's boot screen needs
  // it to answer without the secret to tell a wrong secret from a dead child.
  if (!secret || req.path === '/health') return next();
  // Tried independently rather than `header ?? query`: a wrong or blank header must not shadow the
  // query parameter, which is the only credential EventSource can send.
  if (secretMatches(req.get('X-FastStudy-Secret'), secret)) return next();
  if (secretMatches(req.query.secret, secret)) return next();
  // Chromium reports any other MIME on an EventSource as a bare onerror, so a JSON 401 would read
  // as a transport failure rather than an auth one.
  if ((req.get('Accept') ?? '').includes('text/event-stream')) {
    return res.status(401).type('text/event-stream').end();
  }
  res.status(401).json({ error: 'unauthorized' });
}

// Headers for an outbound call to one of our own services, carrying the launch secret when there
// is one. Peers only — sending it to an external lecture host would leak it (services/probe.js).
export function peerHeaders(headers = {}) {
  const secret = process.env.FASTSTUDY_SECRET;
  return secret ? { ...headers, 'X-FastStudy-Secret': secret } : { ...headers };
}

// Joins a path under the per-user writable state root: FASTSTUDY_STATE_DIR when the launcher sets
// one, else `.state/` at the repo root. A pure join — callers create the directories they write to.
export function statePath(...parts) {
  // src/ -> server/ -> downloader/ -> repo root
  const root =
    process.env.FASTSTUDY_STATE_DIR ??
    path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../../..', '.state');
  return path.join(root, ...parts);
}

File: backend/runtime.py
import os
import socket
from pathlib import Path
from secrets import compare_digest
from urllib.parse import parse_qs

import uvicorn
from dotenv import load_dotenv

# Loaded here, not in main.py: every module that reads env at import time imports `runtime` first,
# so this is the only placement that survives an import re-sort.
load_dotenv()

_SECRET_HEADER = b"x-faststudy-secret"
_UNAUTHORIZED = b'{"error": "unauthorized"}'

# This file is `<repo>/backend/runtime.py`, so the repo root is two levels up.
_REPO_ROOT = Path(__file__).resolve().parent.parent


def secret() -> str | None:
    """The launch secret every caller must present. Unset means no enforcement at all,
    which is what dev runs on."""

    return os.environ.get("FASTSTUDY_SECRET") or None


def state_path(*parts) -> Path:
    """The writable state root with `parts` joined onto it: `FASTSTUDY_STATE_DIR` if set,
    else `.state/` at the repo root."""

    # A pure join that deliberately creates nothing: importing a module that merely names a
    # state file must not leave a directory behind, least of all one redirected elsewhere.
    root = os.environ.get("FASTSTUDY_STATE_DIR") or _REPO_ROOT / ".state"
    return Path(root).joinpath(*parts)


class SecretMiddleware:
    """Answers 401 to any request carrying neither the `X-FastStudy-Secret` header nor a
    `secret` query parameter — EventSource is the one caller that cannot set a header."""

    # Pure ASGI, never BaseHTTPMiddleware: that one buffers a StreamingResponse and would stall SSE.
    def __init__(self, app, secret: str):
        self.app = app
        self.secret = secret.encode()

    async def __call__(self, scope, receive, send):
        # /health is the sole exemption, so the launcher can tell a wrong secret from a dead child.
        exempt = scope["type"] != "http" or scope["path"] == "/health"
        if exempt or self._authorized(scope):
            await self.app(scope, receive, send)
            return
        await self._reject(scope, send)

    def _authorized(self, scope) -> bool:
        given = (
            dict(scope["headers"]).get(_SECRET_HEADER)
            or parse_qs(scope["query_string"]).get(b"secret", [b""])[0]
        )
        return compare_digest(given, self.secret)

    async def _reject(self, scope, send) -> None:
        """An SSE request is refused as an empty `text/event-stream`: Chromium reports any other
        MIME on an EventSource as a bare transport error, hiding the auth failure."""

        sse = b"text/event-stream" in dict(scope["headers"]).get(b"accept", b"")
        body = b"" if sse else _UNAUTHORIZED
        ctype = b"text/event-stream" if sse else b"application/json"
        await send(
            {
                "type": "http.response.start",
                "status": 401,
                "headers": [
                    (b"content-type", ctype),
                    (b"content-length", str(len(body)).encode()),
                ],
            }
        )
        await send({"type": "http.response.body", "body": body})


def serve(app, default_port: int) -> None:
    """Serve `app` on loopback, printing `FASTSTUDY_PORT=<port>` so the launcher can read it.
    The socket is bound here because `uvicorn.run(port=0)` offers no way to learn what it bound."""

    port = int(os.environ.get("FASTSTUDY_PORT", default_port))
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.bind(("127.0.0.1", port))
    # Listen before announcing, so a launcher connecting the instant it reads the line is not refused.
    sock.listen()
    print(f"FASTSTUDY_PORT={sock.getsockname()[1]}", flush=True)
    uvicorn.Server(uvicorn.Config(app)).run(sockets=[sock])
